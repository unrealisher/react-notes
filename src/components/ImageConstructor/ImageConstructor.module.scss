.wrapper {
  position: relative;
}

.file {
  box-sizing: border-box;
  width: 100%;
  padding: 2px 10px;
  border-radius: 15px;
}

.button {
  width: 100%;
  margin-top: 12px;
  padding: 5px 10px;
  border-radius: 15px;
  border: 1px solid rgba(218, 220, 224, 0.5);
  background-color: white;

  &:hover,
  &:focus {
    background-color: #fc0;
  }
}

.image {
  width: 100%;
  height: auto;
  object-fit: cover;
  border-radius: 16px;
}

.reset {
  cursor: pointer;
  position: absolute;
  top: -8px;
  left: calc(100% - 18px);
  width: 25px;
  height: 25px;
  background-color: rgb(124, 124, 124);
  border-radius: 50%;
  z-index: 2;
  opacity: 0.6;

  &::before,
  &::after {
    content: "";
    position: absolute;
    top: calc(50% - 1px);
    left: calc(50% - 10px);
    width: 19px;
    height: 2px;
    background-color: white;
  }

  &::before {
    transform: rotate(45deg);
  }

  &::after {
    transform: rotate(-45deg);
  }

  &:hover {
    opacity: 1;
  }

  &:active {
    opacity: 0.8;
  }
}
