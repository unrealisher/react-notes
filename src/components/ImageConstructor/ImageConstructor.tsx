import React, { useRef } from "react";
import cn from "classnames";

import styles from "./ImageConstructor.module.scss";

interface IProps {
  wrapper?: string;
  image: string;
  setImage: Function;
}

export const ImageConstructor = (props: IProps): JSX.Element => {
  const { wrapper, image, setImage } = props;
  const inputRef = useRef<HTMLInputElement>(null);
  return (
    <div className={cn(styles.wrapper, wrapper ? wrapper : undefined)}>
      {image === "" && (
        <>
          <label>
            Введите URL картинки:
            <input className={styles.file} type="text" ref={inputRef} />
          </label>
          <button
            className={styles.button}
            onClick={() => {
              if (inputRef.current) setImage(inputRef.current.value);
            }}
          >
            Загрузить картинку
          </button>
        </>
      )}
      {image !== "" && (
        <>
          <img
            src={image}
            className={styles.image}
            alt="Упс, что-то пошло не так"
          />
          <button className={styles.reset} onClick={() => setImage("")} />
        </>
      )}
    </div>
  );
};

export default ImageConstructor;
