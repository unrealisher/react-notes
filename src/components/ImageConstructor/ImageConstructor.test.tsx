import React from "react";
import { shallow } from "enzyme";
import { ImageConstructor } from "./ImageConstructor";

describe("ImageConstructor", () => {
  const image =
    "https://avatars.mds.yandex.net/get-pdb/1534889/a869b7ef-32da-41e7-8703-326ae860e67f/orig";
  const setImage = (): void => {};
  it("correct render", () => {
    const component = shallow(
      <ImageConstructor image={image} setImage={setImage} />
    );
    expect(component).toMatchSnapshot();
  });
});
