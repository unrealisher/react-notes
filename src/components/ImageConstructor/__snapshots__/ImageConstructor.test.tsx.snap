// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageConstructor correct render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ImageConstructor
    image="https://avatars.mds.yandex.net/get-pdb/1534889/a869b7ef-32da-41e7-8703-326ae860e67f/orig"
    setImage={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        false,
        <React.Fragment>
          <img
            alt="Упс, что-то пошло не так"
            className="image"
            src="https://avatars.mds.yandex.net/get-pdb/1534889/a869b7ef-32da-41e7-8703-326ae860e67f/orig"
          />
          <button
            className="reset"
            onClick={[Function]}
          />
        </React.Fragment>,
      ],
      "className": "wrapper",
    },
    "ref": null,
    "rendered": Array [
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <img
              alt="Упс, что-то пошло не так"
              className="image"
              src="https://avatars.mds.yandex.net/get-pdb/1534889/a869b7ef-32da-41e7-8703-326ae860e67f/orig"
            />,
            <button
              className="reset"
              onClick={[Function]}
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "alt": "Упс, что-то пошло не так",
              "className": "image",
              "src": "https://avatars.mds.yandex.net/get-pdb/1534889/a869b7ef-32da-41e7-8703-326ae860e67f/orig",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "reset",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": "button",
          },
        ],
        "type": Symbol(react.fragment),
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          false,
          <React.Fragment>
            <img
              alt="Упс, что-то пошло не так"
              className="image"
              src="https://avatars.mds.yandex.net/get-pdb/1534889/a869b7ef-32da-41e7-8703-326ae860e67f/orig"
            />
            <button
              className="reset"
              onClick={[Function]}
            />
          </React.Fragment>,
        ],
        "className": "wrapper",
      },
      "ref": null,
      "rendered": Array [
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <img
                alt="Упс, что-то пошло не так"
                className="image"
                src="https://avatars.mds.yandex.net/get-pdb/1534889/a869b7ef-32da-41e7-8703-326ae860e67f/orig"
              />,
              <button
                className="reset"
                onClick={[Function]}
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "Упс, что-то пошло не так",
                "className": "image",
                "src": "https://avatars.mds.yandex.net/get-pdb/1534889/a869b7ef-32da-41e7-8703-326ae860e67f/orig",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "reset",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": "button",
            },
          ],
          "type": Symbol(react.fragment),
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
